components/dialogs/custom-services/CustomService.state.ts

```diff
@@ -0,0 +1,106 @@
+import { Reducer, useReducer } from 'react';
+
+export interface StateField {
+  value: string;
+  success: boolean;
+}
+
+interface InitialState {
+  username: StateField;
+  email: StateField;
+  request: StateField;
+}
+
+export interface ActionField<T = InitialState> {
+  value: string;
+  name: keyof T;
+}
+
+export interface ActionBlur<T = ActionField> {
+  type: 'blur';
+  field: T;
+}
+
+export interface ActionChange<T = ActionField> {
+  type: 'change';
+  field: T;
+}
+
+export interface DispatchEventOptions {
+  type: 'blur' | 'change';
+  fieldName: string;
+  fieldValue: string;
+}
+
+const initialState: InitialState = {
+  username: {
+    value: '',
+    success: false,
+  },
+  email: {
+    value: '',
+    success: false,
+  },
+  request: {
+    value: '',
+    success: false,
+  },
+};
+
+const reducer: Reducer<InitialState, ActionChange | ActionBlur> = (state, action) => {
+  switch (action.type) {
+    case 'blur': {
+      const { field } = action;
+
+      return {
+        ...state,
+        [field.name]: {
+          ...state[field.name],
+          success: field.name === 'email' ? !!field.value && field.value.includes('@') : !!field.value,
```

Comment:
It looks like `success` state for `'email'` could be moved to a separate function. This action will help us to remove duplication and avoid [the WET anti-pattern](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#WET).